src/modules/preprocessor/analyzer.ts

```diff
@@ -0,0 +1,190 @@
+          })
+        )
+      )
+  const declaredNames = new Dict<
+    string,
+    ArrayMap<string, es.ImportDeclaration['specifiers'][number]>
+  >()
+
+  for (const sourceModule of [...topoOrder, entrypointAbsPath]) {
+    const program = programs[sourceModule]
+    moduleDocs[sourceModule] = new Set()
+
+    for (const node of program.body) {
+      if (node.type === 'ExportDefaultDeclaration') {
+        if (!options.allowUndefinedImports) {
+          assert(
+            !moduleDocs[sourceModule].has('default'),
+            "Multiple default exports should've been caught by the parser"
+          )
+          moduleDocs[sourceModule].add('default')
+        }
+        continue
+      } else if (node.type === 'ExportNamedDeclaration') {
+        if (node.declaration) {
+          if (!options.allowUndefinedImports) {
+            const ids = getIdsFromDeclaration(node.declaration)
+            ids.forEach(id => {
+              assert(id !== null, 'Encountered a null identifier!')
+              return moduleDocs[sourceModule].add(id.name)
+            })
+          }
+          continue
+        }
+
+        for (const spec of node.specifiers) {
+          moduleDocs[sourceModule].add(spec.exported.name)
+        }
+
+        if (!node.source) continue
+      } else if (!isModuleDeclaration(node)) {
+        continue
+      }
+
+      const dstModule = getModuleDeclarationSource(node)
+      const dstModuleDocs = moduleDocs[dstModule]
+
+      if (node.type === 'ExportAllDeclaration') {
+        if (!options.allowUndefinedImports) {
+          if (dstModuleDocs.size === 0) throw new UndefinedNamespaceImportError(dstModule, node)
+
+          if (node.exported) {
+            moduleDocs[sourceModule].add(node.exported.name)
+          } else {
+            for (const each of dstModuleDocs) {
+              if (each === 'default') {
+                // ExportAllDeclarations do not implicitly reexport default exports
+                continue
+              }
+
+              moduleDocs[sourceModule].add(each)
+            }
+          }
+        }
+      } else {
+        for (const spec of node.specifiers) {
+          if (spec.type !== 'ExportSpecifier' && isSourceModule(dstModule)) {
+            const declaredName = spec.local.name
+            declaredNames.setdefault(declaredName, new ArrayMap()).add(dstModule, spec)
+          }
+
+          if (!options.allowUndefinedImports) {
+            if (spec.type === 'ImportNamespaceSpecifier') {
+              if (dstModuleDocs.size === 0) throw new UndefinedNamespaceImportError(dstModule, spec)
+              continue
+            }
+
+            const importedName = getImportedName(spec)
+
+            if (!dstModuleDocs.has(importedName)) {
+              if (importedName === 'default') throw new UndefinedDefaultImportError(dstModule, spec)
+              throw new UndefinedImportError(importedName, dstModule, spec)
+            }
+          }
```

Comment:
The outer `if` could be refactored to a guard clause, thus reducing a layer of code.